src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
+     (sql/delete-rows :factsets ["id=?" factset-id]))))
+
+(defn-validated insert-facts-pv-pairs!
+  [factset-id :- s/Int
+   pairs :- (s/either [(s/pair s/Int "path-id" s/Int "value-id")]
+                      #{(s/pair s/Int "path-id" s/Int "value-id")})]
+  (apply sql/insert-records
+         :facts
+         (for [[pid vid] pairs]
+           {:factset_id factset-id :fact_path_id pid :fact_value_id vid})))
+
+(defn existing-paths->ids
+  "Returns a map from path strings to id for each path that's already
+  in the database."
+  [pathstrs]
+  (into {}
+        (for [{:keys [path id]}
+              (apply query-to-vec
+                     (format "select path, id from fact_paths where path %s"
+                             (jdbc/in-clause pathstrs))
+                     pathstrs)]
+          [path id])))
+
+(defn insert-new-paths!
+  "Inserts the path strings into the database and returns a map from
+  the path strings to their database ids."
+  [pathstrs]
+  (zipmap pathstrs
+          (map :id (apply
+                    sql/insert-records :fact_paths
+                    (map (comp facts/path->pathmap facts/string-to-factpath)
+                         pathstrs)))))
+
+(defn existing-value-hashes->ids
+  "Returns a map from value hash to id for each value that's already
+  in the database."
+  [value-hashes]
+  (into {}
+        (for [{:keys [value_hash id]}
+              (apply query-to-vec
+                     (format "SELECT value_hash, id FROM fact_values
+                                WHERE value_hash %s"
+                             (jdbc/in-clause value-hashes))
+                     value-hashes)]
+          [value_hash id])))
+
+(defn insert-new-values!
+  "Inserts the valuemaps into the database and returns a map from
+  value hash to valuemap.  The returned valuemaps will have an :id."
+  [valuemaps]
+  (let [vhashes (map :value_hash valuemaps)]
+    (zipmap vhashes
+            (let [vmaps valuemaps]
+              (map (fn [id vmap] (assoc vmap :id id))
+                   (map :id (apply sql/insert-records :fact_values valuemaps))
+                   vmaps)))))
+
+(defn realize-paths!
+  [pathstrs]
+  "Ensures that all paths exist in the database and returns a map of
+  paths to ids."
+  (if-let [pathstrs (seq pathstrs)]
+    (let [existing-path-ids (existing-paths->ids pathstrs)
+          missing-db-paths (set/difference (set pathstrs)
+                                           (set (keys existing-path-ids)))]
+      (merge existing-path-ids (insert-new-paths! missing-db-paths)))
+    {}))
+
+(defn realize-values!
+  [valuemaps]
+  "Ensures that all valuemaps exist in the database and returns a
+  map of value hashes to valuemaps with ids."
+  (if-let [valuemaps (seq valuemaps)]
+    (let [value-hashes (map :value_hash valuemaps)
+          existing-vhash-ids (existing-value-hashes->ids value-hashes)
+          vhashes-to-valuemaps (into {} (for [vmap valuemaps]
```

Comment:
I'm a little confused by this whole let block. Seems like we're doing a lot of extra work. Fundamentally we've a list of valuemaps, and we want a map of hash->valuemap, including an existing or newly created id. Here we're adding that id to existing ones, passing through ones without. Even though we know which ones are existing already, on 941 below, we compute it again. Then we take the recomputed value and get the valuemap that we have above on line 943. After that we are overwriting the valuemaps for new ones we passed through here, with the new ids on 945/946.